File: python/text_normalizer.py
import re

class TextNormalizer:
    """
    A class for normalizing text containing unusual writing styles,
    such as repeated characters, text in asterisks, or
    unusual capitalization.
    """
    
    def normalize_text(self, input_text):
        """
        Normalizes text by applying multiple normalization rules.
        
        Args:
            input_text (str): The text to normalize.
        
        Returns:
            str: The normalized text.
        """
        if not input_text:
            return input_text
            
        result = input_text
        
        # Normalize stretched words (e.g., "hellooo" -> "hello")
        result = self.normalize_repeated_characters(result)
        
        # Remove text in asterisks (e.g., "*cough*")
        result = self.remove_asterisks(result)
        
        # Normalize capitalization (e.g., "HeLLo" -> "Hello")
        result = self.normalize_casing(result)
        
        return result
    
    def normalize_repeated_characters(self, input_text):
        """
        Normalizes repeated characters in text.
        
        Args:
            input_text (str): The text to normalize.
        
        Returns:
            str: The text with normalized repeated characters.
        """
        if not input_text:
            return input_text
            
        def normalize_word(match):
            word = match.group(0)
            normalized = []
            i = 0
            
            while i < len(word):
                current_char = word[i]
                normalized.append(current_char)
                
                # Skip all subsequent identical characters
                while i + 1 < len(word) and word[i + 1] == current_char:
                    i += 1
                
                i += 1
                
            return ''.join(normalized)
        
        # Process each word (sequence of word characters)
        return re.sub(r'\b\w+\b', normalize_word, input_text)
    
    def remove_asterisks(self, input_text):
        """
        Removes text surrounded by asterisks and fixes any resulting extra spaces.
        
        Args:
            input_text (str): The text to normalize.
        
        Returns:
            str: The text without asterisk content and no extra spaces.
        """
        if not input_text:
            return input_text
            
        # Remove text between asterisks completely
        result = re.sub(r'\*[^*]*\*', '', input_text)
        
        # Fix any resulting double spaces or more
        result = re.sub(r'\s{2,}', ' ', result)
        
        return result
    
    def normalize_casing(self, input_text):
        """
        Normalizes capitalization of words.
        
        Args:
            input_text (str): The text to normalize.
        
        Returns:
            str: The text with normalized capitalization.
        """
        if not input_text:
            return input_text
            
        def normalize_word_casing(match):
            word = match.group(0)
            
            # If the word has mixed capitalization
            if self._has_mixed_casing(word):
                # Convert everything to lowercase
                normalized = word.lower()
                
                # First letter uppercase if the original was uppercase
                if word[0].isupper():
                    normalized = normalized[0].upper() + normalized[1:]
                
                return normalized
            
            return word
            
        # Process each word (sequence of word characters)
        return re.sub(r'\b\w+\b', normalize_word_casing, input_text)
    
    def _has_mixed_casing(self, word):
        """
        Checks if a word contains mixed capitalization.
        
        Args:
            word (str): The word to check.
        
        Returns:
            bool: True if the word has mixed capitalization, otherwise False.
        """
        if len(word) <= 1:
            return False
            
        has_upper = False
        has_lower = False
        
        for char in word:
            if char.isupper():
                has_upper = True
            elif char.islower():
                has_lower = True
                
            if has_upper and has_lower:
                return True
                
        return False


# Example usage
if __name__ == "__main__":
    normalizer = TextNormalizer()
    
    # Example with stretched words and asterisks
    sample_text = "Hellooo how areee you *cough* today? It's a niiice day *laughs*"
    normalized = normalizer.normalize_text(sample_text)
    print(f"Original: {sample_text}")
    print(f"Normalized: {normalized}")
    
    # Example with mixed capitalization
    sample_text2 = "HeLLo ThERe, HOw Are YOu?"
    normalized2 = normalizer.normalize_text(sample_text2)
    print(f"Original: {sample_text2}")
    print(f"Normalized: {normalized2}")

File: rust/text_normalizer.rs
use regex::Regex;
use std::borrow::Cow;

/// A struct for normalizing text containing unusual writing styles,
/// such as repeated characters, text in asterisks, or
/// unusual capitalization.
pub struct TextNormalizer;

impl TextNormalizer {
    /// Creates a new instance of TextNormalizer.
    pub fn new() -> Self {
        TextNormalizer
    }

    /// Normalizes text by applying multiple normalization rules.
    ///
    /// # Arguments
    ///
    /// * `input` - The text to normalize.
    ///
    /// # Returns
    ///
    /// The normalized text.
    pub fn normalize_text<'a>(&self, input: &'a str) -> Cow<'a, str> {
        if input.is_empty() {
            return Cow::Borrowed(input);
        }

        let result = self.normalize_repeated_characters(input);
        let result = self.remove_asterisks(&result);
        let result = self.normalize_casing(&result);

        result
    }

    /// Normalizes repeated characters in text.
    ///
    /// # Arguments
    ///
    /// * `input` - The text to normalize.
    ///
    /// # Returns
    ///
    /// The text with normalized repeated characters.
    pub fn normalize_repeated_characters<'a>(&self, input: &'a str) -> Cow<'a, str> {
        if input.is_empty() {
            return Cow::Borrowed(input);
        }

        // Define word boundaries to identify words
        let re = Regex::new(r"\b\w+\b").unwrap();
        
        let result = re.replace_all(input, |caps: &regex::Captures| {
            let word = &caps[0];
            let mut normalized = String::with_capacity(word.len());
            let mut chars = word.chars().peekable();
            
            while let Some(current) = chars.next() {
                normalized.push(current);
                
                // Skip all subsequent identical characters
                while chars.peek() == Some(&current) {
                    chars.next();
                }
            }
            
            normalized
        });

        result
    }

    /// Removes text surrounded by asterisks and fixes any resulting extra spaces.
    ///
    /// # Arguments
    ///
    /// * `input` - The text to normalize.
    ///
    /// # Returns
    ///
    /// The text without asterisk content and no extra spaces.
    pub fn remove_asterisks<'a>(&self, input: &'a str) -> Cow<'a, str> {
        if input.is_empty() {
            return Cow::Borrowed(input);
        }
        
        // Remove text between asterisks completely
        let re = Regex::new(r"\*[^*]*\*").unwrap();
        let result = re.replace_all(input, "");
        
        // Fix any resulting double spaces or more
        let space_re = Regex::new(r"\s{2,}").unwrap();
        let result = space_re.replace_all(&result, " ");
        
        result
    }

    /// Normalizes capitalization of words.
    ///
    /// # Arguments
    ///
    /// * `input` - The text to normalize.
    ///
    /// # Returns
    ///
    /// The text with normalized capitalization.
    pub fn normalize_casing<'a>(&self, input: &'a str) -> Cow<'a, str> {
        if input.is_empty() {
            return Cow::Borrowed(input);
        }

        // Define word boundaries to identify words
        let re = Regex::new(r"\b\w+\b").unwrap();
        
        let result = re.replace_all(input, |caps: &regex::Captures| {
            let word = &caps[0];
            
            // If the word has mixed capitalization
            if self.has_mixed_casing(word) {
                // Convert everything to lowercase
                let normalized = word.to_lowercase();
                
                // First letter uppercase if the original was uppercase
                if let Some(first_char) = word.chars().next() {
                    if first_char.is_uppercase() {
                        let mut chars: Vec<char> = normalized.chars().collect();
                        if !chars.is_empty() {
                            chars[0] = chars[0].to_uppercase().next().unwrap_or(chars[0]);
                        }
                        return chars.iter().collect::<String>();
                    }
                }
                
                normalized
            } else {
                word.to_string()
            }
        });
        
        result
    }

    /// Checks if a word contains mixed capitalization.
    ///
    /// # Arguments
    ///
    /// * `word` - The word to check.
    ///
    /// # Returns
    ///
    /// True if the word has mixed capitalization, otherwise False.
    fn has_mixed_casing(&self, word: &str) -> bool {
        if word.len() <= 1 {
            return false;
        }
        
        let mut has_upper = false;
        let mut has_lower = false;
        
        for c in word.chars() {
            if c.is_uppercase() {
                has_upper = true;
            } else if c.is_lowercase() {
                has_lower = true;
            }
            
            if has_upper && has_lower {
                return true;
            }
        }
        
        false
    }
}

impl Default for TextNormalizer {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_text() {
        let normalizer = TextNormalizer::new();
        
        assert_eq!(
            normalizer.normalize_text("Hellooo how areee you *cough* today?"),
            "Hello how are you today?"
        );
        
        assert_eq!(
            normalizer.normalize_text("HeLLo ThERe, HOw Are YOu?"),
            "Hello There, How Are You?"
        );
        
        assert_eq!(
            normalizer.normalize_text("It's a niiice *laughs* day!"),
            "It's a nice day!"
        );
        
        assert_eq!(normalizer.normalize_text(""), "");
        
        assert_eq!(
            normalizer.normalize_text("Regular text with no issues."),
            "Regular text with no issues."
        );
    }

    #[test]
    fn test_normalize_repeated_characters() {
        let normalizer = TextNormalizer::new();
        
        assert_eq!(
            normalizer.normalize_repeated_characters("Hellooo"),
            "Hello"
        );
        
        assert_eq!(
            normalizer.normalize_repeated_characters("goooood moooorning"),
            "good morning"
        );
        
        assert_eq!(
            normalizer.normalize_repeated_characters("normal text"),
            "normal text"
        );
        
        assert_eq!(normalizer.normalize_repeated_characters(""), "");
    }

    #[test]
    fn test_remove_asterisks() {
        let normalizer = TextNormalizer::new();
        
        assert_eq!(
            normalizer.remove_asterisks("Hello *cough* there"),
            "Hello there"
        );
        
        assert_eq!(
            normalizer.remove_asterisks("*laughs* This is funny *smiles*"),
            "This is funny"
        );
        
        assert_eq!(
            normalizer.remove_asterisks("No asterisks here"),
            "No asterisks here"
        );
        
        assert_eq!(normalizer.remove_asterisks(""), "");
    }

    #[test]
    fn test_normalize_casing() {
        let normalizer = TextNormalizer::new();
        
        assert_eq!(normalizer.normalize_casing("HeLLo"), "Hello");
        
        assert_eq!(normalizer.normalize_casing("ALL CAPS"), "ALL CAPS");
        
        assert_eq!(
            normalizer.normalize_casing("all lowercase"),
            "all lowercase"
        );
        
        assert_eq!(
            normalizer.normalize_casing("MiXeD CaSiNg TeXt"),
            "Mixed Casing Text"
        );
        
        assert_eq!(normalizer.normalize_casing(""), "");
    }

    #[test]
    fn test_has_mixed_casing() {
        let normalizer = TextNormalizer::new();
        
        assert!(normalizer.has_mixed_casing("HeLLo"));
        assert!(!normalizer.has_mixed_casing("HELLO"));
        assert!(!normalizer.has_mixed_casing("hello"));
        assert!(!normalizer.has_mixed_casing("Hello")); // First capital letter only is not mixed case
        assert!(!normalizer.has_mixed_casing("a"));     // Single character
        assert!(!normalizer.has_mixed_casing(""));      // Empty string
    }
}